### PR TITLE
Add clang-17 to CI

### LINKED ISSUE #2169: Add support for clang-17
LLVM 17 was just released: https://discourse.llvm.org/t/llvm-17-0-1-released/73549

Let's add support for clang-17 to the CI.

### PR DESCRIPTION
**Pull request by bernhardmgruber**
Fixes: #2169

### CODE REVIEW

**Comment by j-stephan** [comment]
Okay, this is definitely the last feature PR going into 1.0. If some other compiler releases today or tomorrow it will not be included, nor will any fixes for this compiler.

I'll submit a PR which cherry-picks the commits since the creation of the 1.0 branch; afterwards we should be ready for the release.

**Comment by bernhardmgruber** [comment]
> Okay, this is definitely the last feature PR going into 1.0. If some other compiler releases today or tomorrow it will not be included, nor will any fixes for this compiler.

I understand :)

**Comment by SimeonEhrig** [comment]
Clang 17 is not official supported by the `alpaka-job-coverage-library`. Please open an issue: https://github.com/alpaka-group/alpaka-job-matrix-library/issues

**Comment by bernhardmgruber** [comment]
> Clang 17 is not official supported by the `alpaka-job-coverage-library`. Please open an issue: https://github.com/alpaka-group/alpaka-job-matrix-library/issues

Wait. Now adding a new compiler version also requires to add issues, PRs and commits to other repositories? That seems a bit .. complex.

Anyhow, here is the issue: https://github.com/alpaka-group/alpaka-job-matrix-library/issues/39

**Comment by SimeonEhrig** [comment]
*

> > Clang 17 is not official supported by the `alpaka-job-coverage-library`. Please open an issue: https://github.com/alpaka-group/alpaka-job-matrix-library/issues
> 
> Wait. Now adding a new compiler version also requires to add issues, PRs and commits to other repositories? That seems a bit .. complex.
> 
> Anyhow, here is the issue: [alpaka-group/alpaka-job-matrix-library#39](https://github.com/alpaka-group/alpaka-job-matrix-library/issues/39)

Yes, but only because we share the CI with cupla, vikunja and PIConGPU.

**Comment by bernhardmgruber** [comment]
I am moving this off to after the 1.0 release, since the remaining failures are with `-Werror` and CUDA mode. alpaka compiles fine with clang-17 outside those use cases.

**Comment by bernhardmgruber** [comment]
There are failures with generating ptx again:
```
ptxas /tmp/randomStrategies-sm_61-768a17.s, line 14415; fatal   : Parsing error near '.': syntax error
ptxas fatal   : Ptx assembly aborted due to errors
clang++: error: ptxas command failed with exit code 255 (use -v to see invocation)
```
I think we had this before somewhere?

**Comment by j-stephan** [comment]
Yes, it is this one: https://github.com/alpaka-group/alpaka/issues/1906 Glad to see this issue persisting on LLVM's side...

**Comment by bernhardmgruber** [comment]
> Yes, it is this one: #1906 Glad to see this issue persisting on LLVM's side...

Thx, I excluded clang-17 as CUDA compiler in Debug mode, as we do for older clang versions.

**Comment by j-stephan** [comment]
You also need to update `README.md` :-)

**Comment by bernhardmgruber** [comment]
> You also need to update `README.md` :-)

Right. Added clang-17 for all CPU backends, because I did not have the chance to test what combinations of nvcc+clang and clang CUDA work.

File: README.md [hunk 1]
```diff
@@ -65,17 +65,17 @@ Supported Compilers
...
+| CUDA (clang)                                                                   | -                                               | -                                               | -                                           | :x:                    | :x:                                                        | :x:                                                   | :x:                                             | :x:                                         | :white_check_mark: (CUDA 11.0 - 11.5)             | :white_check_mark: (CUDA 11.0 - 11.5)[^1] | :white_check_mark: (CUDA 11.0 - 11.5)[^1] | :x:                     | :x:                              | -                                                     | -                                    |
```
Comment (SimeonEhrig) [inline README.md]: Why is CUDA (clang) not supported? Clang 17 supports up to CUDA 12.1: https://llvm.org/docs/CompileCudaWithLLVM.html
And we have a CI job, which test it.

Comment (bernhardmgruber) [reply README.md]: I added the range CUDA 11.0 - 12.1 then.

**Review by SimeonEhrig — DISMISSED** [review]

File: README.md (hunk 1 above)
Comment (j-stephan) [reply README.md]: According to https://github.com/llvm/llvm-project/blob/llvmorg-17.0.0/clang/include/clang/Basic/Cuda.h 11.8 is the last "fully supported" version, 12.1 is "partially supported". So far we always used the "fully supported" version in our matrix.

Comment (bernhardmgruber) [reply README.md]: So CUDA 11.0 - 11.8 then?

Comment (j-stephan) [reply README.md]: If we want to stay consistent, yes.

Comment (bernhardmgruber) [reply README.md]: Alright.

**Review by j-stephan — APPROVED** [review]

**Pull request merged**